src/Disks/IO/AsynchronousBoundedReadBuffer.cpp

```diff
@@ -132,7 +145,7 @@ void AsynchronousBoundedReadBuffer::prefetch(Priority priority)
     last_prefetch_info.submit_time = std::chrono::system_clock::now();
     last_prefetch_info.priority = priority;
 
-    prefetch_buffer.resize(buffer_size);
+    prefetch_buffer.resize(calculateBufferSizeForNextRead());
```

Comment:
It is not obvious, why it actually helps since resize for smaller size doesn't seem to release any memory:

https://github.com/ClickHouse/ClickHouse/blob/4e6d59478241b0ecd702932964a459de37c0d865/src/IO/BufferWithOwnMemory.h#L87-L91

could you elaborate on that pls, maybe deserves a comment in code

Reply:
The fix is all about the first allocation.
The problematic scenario is when we have 1000 small columns, each one much smaller than buffer_size. 
Previously first calls to `prefetch_buffer.resize(buffer_size)` for each column will allocate in total 1000*buffer_size of memory which is several GBs.
With this change the first calls will allocate just around the total size of all columns.

Reply:
It should not, since `AsynchronousBoundedReadBuffer` should be created with `buffer_size` = size to read, via `MergeTreeReaderStreamMultipleColumns::estimateMarkRangeBytesOneColumn`

I believe that the problem with system.metric_log is different, it is wide parts + too huge `max_insert_delayed_streams_for_parallel_write`

Hence I have two fixes:
- https://github.com/ClickHouse/ClickHouse/pull/77919
- https://github.com/ClickHouse/ClickHouse/pull/77922

And after any of this change the memory consumption for system.metric_log significantly reduced for me (i.e. with `max_insert_delayed_streams_for_parallel_write=1` it is few megabytes - 20, with `max_insert_delayed_streams_for_parallel_write=1000` it is ~5GiB, for just few rows, where few < 1000)

Reply:
I was able to finally figure out why we see `prefetch` in stacktraces, it should be fixed by https://github.com/ClickHouse/ClickHouse/pull/77993